hack/lib/golang.sh

```diff
@@ -467,12 +467,17 @@ kube::golang::verify_go_version() {
   # default GO_VERSION to content of .go-version
   GO_VERSION="${GO_VERSION:-"$(cat "${KUBE_ROOT}/.go-version")"}"
   # only setup go if we haven't set FORCE_HOST_GO, or `go version` doesn't match GO_VERSION
-  if ! ([ -n "${FORCE_HOST_GO:-}" ] || \
+  # and GOTOOLCHAIN is not set (the forward thinking idiomatic alternative FORCE_HOST_GO)
+  if [ -n "${FORCE_HOST_GO:-}" ]; then
+    GOTOOLCHAIN="${GOTOOLCHAIN:-local}"
+    export GOTOOLCHAIN
+  elif ! ([ -n "${GOTOOLCHAIN:-}" ] || \
```

Comment:
wondering if we'll encounter things explicitly setting GOTOOLCHAIN to its default ("GOTOOLCHAIN=auto") and if we should make this only skip our go version setup if a non-default GOTOOLCHAIN value is detected

Reply:
yeah, https://github.com/kubernetes/kubernetes/pull/120279#issuecomment-1699824195 raises that possibility.

+1 to make "auto" behave like unset. I'll implement that

Reply:
This is a nit, but mixing if / elif and `||` block chaining is hard to follow... is there a clearer way to see the cases we're handling.

Also, I think we're missing a case where 1) GOTOOLCHAIN was not set, 2) GO_VERSION was set, 3) _go_ already exists and matches GO_VERSION. Shouldn't we be setting GOTOOLCHAIN in that case?

```
if FORCE_HOST_GO:
  export GOTOOLCHAIN=local
elseif non-empty non-"auto" GOTOOLCHAIN:
  no-op, honor GOTOOLCHAIN
elsif go exists and is already at the right version:
  export GOTOOLCHAIN="go${GO_VERSION}" // missing case
else
  gimme "${GO_VERSION}"
  export GOTOOLCHAIN="go${GO_VERSION}" // missing case
```

Reply:
we should now treat `auto` like unset

Reply:
/hold
I missed https://github.com/kubernetes/kubernetes/pull/120279#discussion_r1321893169 in the prior updates.
I agree this is worth refactoring.